### PR TITLE
616. Add Bold Tag in String, using a boolean array.

### PR DESCRIPTION
**Pull request by altayhunoglu**
Refer to: https://github.com/spiralgo/algorithms/issues/72

 In the first place, we mark all characters which will be bold.
 We use a boolean array to mark the characters of the string (**s**).
 A character should be marked if it is a char of any substring in the dictionary (**dict**).

(Note that, although the dictionary does not contain duplicates, any substring in the dictionary may occur multiple times in the string 's' )

After the marking process, it is quite easy to wrap marked parts of the string using a StringBuilder.

image

### CODE REVIEW

File: src/main/java/algorithms/curated170/medium/AddBoldTaginString.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+                if(!startB){
+                    startB = true;
+                    sb.append("<b>");
+                }
+            }else{
+                if(startB){
+                    startB = false;
+                    sb.append("</b>");
+                }
+            }
```
Comment (ErdemT09) [inline src/main/java/algorithms/curated170/medium/AddBoldTaginString.java:30]: These are equivalent to 
```
if(arr[i] && !startB){
                    startB = true;
                    sb.append("<b>");
            }

else if(startB){
                    startB = false;
                    sb.append("</b>");
            }
```

File: src/main/java/algorithms/curated170/medium/AddBoldTaginString.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        StringBuilder sb = new StringBuilder();
+        boolean arr[] = new boolean[s.length()];
+        for(String d : dict){
+            int start = s.indexOf(d);
+            while(start != -1){
+                int length = start+ d.length();
+                for(int i = start; i < length; i++){
+                    arr[i] = true;
+                }
+                start = s.indexOf(d, start + 1);
```
Comment (ErdemT09) [inline src/main/java/algorithms/curated170/medium/AddBoldTaginString.java:15]: Is this used to check the recurrences of the same string `d` inside of the string `s` after its previously detected index?

File: src/main/java/algorithms/curated170/medium/AddBoldTaginString.java (hunk 1 above)
Comment (altayhunoglu) [reply src/main/java/algorithms/curated170/medium/AddBoldTaginString.java:30]: Yeah, you are right. I just thought the first one was more readable. But, we can alter the code. No problem.

Comment (altayhunoglu) [reply src/main/java/algorithms/curated170/medium/AddBoldTaginString.java]: Oke, we fixed it like this.

```
  if(arr[i] && !startB){
                    startB = true;
                    sb.append("<b>");
             }else if(!arr[i] && startB){
                    startB = false;
                    sb.append("</b>");
                }
            
```

**Review by ErdemT09 — APPROVED** [review]
Notable for its usage of StringBuilder and its methods.  Using a boolean array to keep track of some values reminded me of [Counting Primes](https://github.com/spiralgo/algorithms/blob/master/src/main/java/algorithms/easy/CountPrimes.java) and [Subset Sum ](https://github.com/spiralgo/algorithms/blob/master/src/main/java/algorithms/SubsetSum.java#L48).

**Pull request merged**